Wangscape/noise/module/Bitmap.cpp

```diff
@@ -0,0 +1,116 @@
+#include "Bitmap.h"
+#include "logging/Logging.h"
+
+namespace noise
+{
+namespace module
+{
+
+namespace {
+// TODO (Serin): extract the data format from noise::RasterBase into a separate class and use it here
+static std::map<std::string, std::shared_ptr<sf::Image>> loadedImages;
+}
+
+
+Bitmap::Bitmap() : 
+    Module::Module(GetSourceModuleCount())
+{
+}
+
+const std::string & Bitmap::GetFilename() const
+{
+    return mFilename;
+}
+
+void Bitmap::SetFilename(const std::string & filename)
+{
+    // Load and process file if not already done
+    auto it = loadedImages.find(filename);
+    if (it == loadedImages.cend())
+        return;
+    else
```

Comment:
Yes 😅. When there are two decisions "on the same level", prime/composite number for example.

I used to implement a "main body" of a function under `if`, which checked a positive variant, leaving error handling for the rest of a function, but next I've switch to checking preconditions at the beginning as you've done here and that's a better way I think.